File: 2024/day_01.py
import argparse

from collections import Counter
from pathlib import Path
from time import time

def _parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--part", "-p",
        type=int,
        choices={1, 2},
        help="Set puzzle part"
    )
    args = parser.parse_args()
    if not args.part:
        parser.error("Which part are you solving?")
    return args

if __name__ == "__main__":
    args = _parse_args()
    t = time()
    with Path(f"{Path(__file__).parent}/inputs/{Path(__file__).stem}.txt").open("r") as file:
        data = file.read().split("\n")
    left = [int(row.split("   ")[0]) for row in data]
    right = [int(row.split("   ")[1]) for row in data ]
    if args.part == 1:
        left.sort()
        right.sort()
        print(sum([abs(l - r) for l, r in zip(left, right)]))
    else:
        right_count = Counter(right)
        similarity = 0
        for l in left:
            similarity += l * right_count.get(l, 0)
        print(similarity)
    print(time() - t)

File: 2021/day_01.rb
require 'optparse'
require 'ostruct'

def parse_args
  args = OpenStruct.new
  OptionParser.new do |arg|
    arg.on('-p', '--part PART', 'The part you\'re solving') { |a| args.part = a }
  end.parse!
  if not args.part
    raise "Which part are you solving?"
  end
  return args
end

def count_increases(data)
  counter = 0
  data.each_with_index {
    |val, idx|
    if idx > 0 and val > data[idx - 1]
      counter += 1
    end
  }
  return counter
end

if __FILE__ == $0
    args = parse_args
    t = Time.now
    data = nil
    File.open("#{File.dirname($0)}/inputs/#{File.basename($0, ".*")}.txt", "r") do |f|
      data = f.read.split("\n").map { |x| x.to_i }
    end

    if args.part == "2"
      data = (0..data.length - 3).map {
        |idx|
        data[idx] + data[idx + 1] + data[idx + 2]
      }
    end

    puts count_increases(data)
    puts Time.now - t
end
